query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs

```diff
@@ -63,113 +62,75 @@ impl ChainingHashTable {
-    }
 
-    fn compute_hash_table_size(rows_count: usize) -> usize {
-        // The next power of 2
-        let mut power = 1;
-        while power < rows_count {
-            power <<= 1;
+        for col in keys {
+            Series::serialize(col, &mut serialized_keys, None)?
         }
-        power
+        Ok(serialized_keys)
     }
 }
 
 impl HashJoinState for ChainingHashTable {
     fn build(&self, input: DataBlock) -> Result<()> {
+        let func_ctx = self.ctx.try_get_function_context()?;
         let build_keys = self
             .build_expressions
             .iter()
-            .map(|expr| {
-                ExpressionEvaluator::eval(self.ctx.try_get_function_context()?, expr, &input)
-            })
+            .map(|expr| ExpressionEvaluator::eval(&func_ctx, expr, &input))
             .collect::<Result<Vec<ColumnRef>>>()?;
 
-        let hash_values = self.hash(&build_keys, input.num_rows())?;
-
-        self.row_space.push(input, hash_values)?;
+        let serialized_build_keys = self.serialize_keys(&build_keys, input.num_rows())?;
+        self.row_space.push(input, serialized_build_keys)?;
 
         Ok(())
     }
 
     fn probe(&self, input: &DataBlock) -> Result<Vec<DataBlock>> {
+        let func_ctx = self.ctx.try_get_function_context()?;
         let probe_keys = self
             .probe_expressions
             .iter()
-            .map(|expr| {
-                ExpressionEvaluator::eval(self.ctx.try_get_function_context()?, expr, input)
-            })
+            .map(|expr| ExpressionEvaluator::eval(&func_ctx, expr, input))
             .collect::<Result<Vec<ColumnRef>>>()?;
-
+        let serialized_probe_keys = self.serialize_keys(&probe_keys, input.num_rows())?;
         let hash_table = self.hash_table.read().unwrap();
-        let hash_values = self.hash(&probe_keys, input.num_rows())?;
-        let hash_values =
-            ChainingHashTable::apply_bitmask(&hash_values, (hash_table.len() - 1) as u64);
 
         let mut results: Vec<DataBlock> = vec![];
-        for (i, hash_value) in hash_values.iter().enumerate().take(input.num_rows()) {
-            let probe_result_ptrs = hash_table[*hash_value as usize].as_slice();
-            if probe_result_ptrs.is_empty() {
+        for (i, key) in serialized_probe_keys
+            .iter()
+            .enumerate()
+            .take(input.num_rows())
+        {
+            let keys_ref = KeysRef::create(key.as_ptr() as usize, key.len());
+            let probe_result_ptr = hash_table.find_key(&keys_ref);
+            if probe_result_ptr.is_none() {
                 // No matched row for current probe row
                 continue;
             }
+            let probe_result_ptrs = probe_result_ptr.unwrap().get_value();
+            // `result_block` is the block of build table
             let result_block = self.row_space.gather(probe_result_ptrs)?;
+            let mut probe_block = DataBlock::block_take_by_indices(input, &[i as u32])?;
```

Comment:
`result_block` and `probe_block` seems have different size.

Reply:
Yes, but I don't get your concerned point in the comment.

Reply:
so `probe_block.add_column(col.clone(), field.clone())` may be incorrect if they are in different size.

Reply:
got it

Reply:
we should use `replicated` to expand the probe_bock to the same size as result_block;

The probe_block may not be in just one size.

For example:

```
 select count(*) from (select 1 as a from t1 order by a limit 10) t1 inner join (select 1 as a from t2 order by a limit 10) t2 on t1.a=t2.a
```
Expect to have 100 output, but just 10;

Reply:
```
duckdb> select count(*) from (select 1 as a from range(0,10)) t1 inner join (select 1 as a from range(0, 10)) t2 on t1.a=t2.a;
┌──────────────┐
│ count_star() │
╞══════════════╡
│          100 │
└──────────────┘
Elapsed: 4 ms

duckdb> 
```

Reply:
> Expect to have 100 output, but actually it's 10;

I tested in my local, the output is 100.
```sql
mysql>  select count(*) from (select 1 as a from t1 order by a limit 10) t1 inner join (select 1 as a from t2 order by a limit 10) t2 on t1.a=t2.a
    -> ;
+----------+
| COUNT(*) |
+----------+
|      100 |
+----------+
1 row in set (9.82 sec)
Read 30000000 rows, 228.88 MiB in 9.789 sec., 3.06 million rows/sec., 23.38 MiB/sec.

```

Reply:
That's interesting
```
:) select count(*) from (select 1 as a from t1 order by a limit 10) t1 inner join (select 1 as a from t2 order by a limit 10) t2 on t1.a=t2.a;
+----------+
| COUNT(*) |
+----------+
|       10 |
+----------+
1 row in set (0.67 sec)
Read 30000000 rows, 228.88 MiB in 0.666 sec., 45.04 million rows/sec., 343.63 MiB/sec.
```

Reply:
t1 and t2 are the following?
```
create table t1(a int, b int);
create table t2(a int, b int);

insert into t1 select number, number+1 from numbers(100000000);
insert into t2 select number, number+1 from numbers(20000000);
```

Reply:
Oh, with your latest commit, it's ok now